### PR TITLE
Cleanup temporary files created by functions

### PR DESCRIPTION
**Pull request by jerrypeng**
### Motivation

Currently, in functions internal code, temporary files are created for purposes of downloading function packages locally.  These files care never cleaned up and may linger around for a long time.  In the best case, this will only cause storage to be unnecessarily wasted but in the worst case this can cause memory issues (which usually is going to be a lot smaller than storage) especially for brokers/workers running in containers since the whole filesystem is just in memory. 

### Modifications

Clean up the temporary files that are created.  To do this I also need to do the following:

1. Refactor the code because the temporary file creation was deeply embedded and sprinkled in random places

2. Because I need to refactor the code, I also needed separate out the code for V3 endpoints and the code for V2 endpoints as all that logic is currently mushed together and extremely convoluted.

3. Fix all the tests because of the refactoring

4. After the refactoring, the temporary files can be referenced thus deleted.

5. Added tests to test temp file cleanup

### CODE REVIEW

File: pulsar-broker/src/test/java/org/apache/pulsar/io/PulsarFunctionE2ETest.java [hunk 1]
```diff
@@ -450,6 +452,15 @@ private void testE2EPulsarFunction(String jarFilePathUrl) throws Exception {
         // make sure subscriptions are cleanup
         assertEquals(admin.topics().getStats(sourceTopic).subscriptions.size(), 0);
 
+        // make sure all temp files are deleted
+        File dir = new File(System.getProperty("java.io.tmpdir"));
```
Comment (merlimat) [inline pulsar-broker/src/test/java/org/apache/pulsar/io/PulsarFunctionE2ETest.java]: Maybe we could have a configurable directory for these download & test operation. Given that /tmp is most commonly mounted on tmpfs which is a ramdisk, we should allow users to configure a different directory that is disk-based (eg: using emptyDir on K8S)

Comment (jerrypeng) [reply pulsar-broker/src/test/java/org/apache/pulsar/io/PulsarFunctionE2ETest.java]: @merlimat yup I agree.  That is why I have consolidated the code that creates these temp files.  This PR is already sizable so lets do this in a subsequent PR

**Comment by jerrypeng** [comment]
rerun cpp tests

**Comment by jerrypeng** [comment]
rerun cpp tests

**Comment by jerrypeng** [comment]
Please let me merge this PR myself after I do some final testing.  Thanks!

**Comment by jerrypeng** [comment]
rerun java8 tests
rerun integration tests

**Comment by jerrypeng** [comment]
rerun java8 tests

**Comment by jerrypeng** [comment]
rerun java8 tests

**Comment by jerrypeng** [comment]
rerun java8 tests

**Pull request merged**